File: Edu-JavaScript/src/com/jetbrains/edu/javascript/learning/JsCourseBuilder.java
package com.jetbrains.edu.javascript.learning;

import com.jetbrains.edu.learning.EduCourseBuilder;
import com.jetbrains.edu.learning.LanguageSettings;
import com.jetbrains.edu.learning.courseFormat.Course;
import com.jetbrains.edu.learning.newproject.CourseProjectGenerator;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class JsCourseBuilder implements EduCourseBuilder<JsNewProjectSettings> {
  @Nullable
  @Override
  public String getTaskTemplateName() {
    return "task.js";
  }

  @Nullable
  @Override
  public String getTestTemplateName() {
    return "test.js";
  }

  @NotNull
  @Override
  public LanguageSettings<JsNewProjectSettings> getLanguageSettings() {
    return new JsLanguageSettings();
  }

  @Nullable
  @Override
  public CourseProjectGenerator<JsNewProjectSettings> getCourseProjectGenerator(@NotNull Course course) {
    return new JsCourseProjectGenerator(this, course);
  }
}

File: Edu-JavaScript/src/com/jetbrains/edu/javascript/learning/JsCourseBuilder.kt
package com.jetbrains.edu.javascript.learning

import com.intellij.lang.javascript.ui.NodeModuleNamesUtil.PACKAGE_JSON
import com.intellij.openapi.project.Project
import com.jetbrains.edu.learning.*
import com.jetbrains.edu.learning.courseFormat.Course
import com.jetbrains.edu.learning.courseFormat.ext.getDir
import com.jetbrains.edu.learning.newproject.CourseProjectGenerator

class JsCourseBuilder : EduCourseBuilder<JsNewProjectSettings> {
  override fun taskTemplateName(course: Course): String = JsConfigurator.TASK_JS
  override fun testTemplateName(course: Course): String = JsConfigurator.TEST_JS

  override fun getLanguageSettings(): LanguageSettings<JsNewProjectSettings> = JsLanguageSettings()
  override fun getCourseProjectGenerator(course: Course): CourseProjectGenerator<JsNewProjectSettings> =
    JsCourseProjectGenerator(this, course)

  override fun refreshProject(project: Project, cause: RefreshCause) {
    if (cause == RefreshCause.DEPENDENCIES_UPDATED) {
      val baseDir = project.course?.getDir(project.courseDir) ?: return
      val packageJson = baseDir.findChild(PACKAGE_JSON) ?: return
      installNodeDependencies(project, packageJson)
    }
  }
}
